drivers/SmartThings/matter-thermostat/src/init.lua

```diff
@@ -283,13 +458,166 @@ local function do_configure(driver, device)
+  [capabilities.carbonMonoxideMeasurement.NAME] = units.PPM,
+  [capabilities.carbonDioxideMeasurement.NAME] = units.PPM,
+  [capabilities.nitrogenDioxideMeasurement.NAME] = units.PPM,
+  [capabilities.ozoneMeasurement.NAME] = units.PPM,
+  [capabilities.formaldehydeMeasurement.NAME] = units.PPM,
+  [capabilities.veryFineDustSensor.NAME] = units.UGM3,
+  [capabilities.fineDustSensor.NAME] = units.UGM3,
+  [capabilities.dustSensor.NAME] = units.UGM3,
+  [capabilities.radonMeasurement.NAME] = units.BQM3,
+  [capabilities.tvocMeasurement.NAME] = units.PPM
+}
+
+-- All ConcentrationMesurement clusters inherit from the same base cluster definitions,
+-- so CarbonMonoxideConcentratinMeasurement is used below but the same enum types exist
+-- in all ConcentrationMeasurement clusters
+local level_strings = {
+  [clusters.CarbonMonoxideConcentrationMeasurement.types.LevelValueEnum.UNKNOWN] = "unknown",
+  [clusters.CarbonMonoxideConcentrationMeasurement.types.LevelValueEnum.LOW] = "good",
+  [clusters.CarbonMonoxideConcentrationMeasurement.types.LevelValueEnum.MEDIUM] = "moderate",
+  [clusters.CarbonMonoxideConcentrationMeasurement.types.LevelValueEnum.HIGH] = "unhealthy",
+  [clusters.CarbonMonoxideConcentrationMeasurement.types.LevelValueEnum.CRITICAL] = "hazardous",
+}
+
+local molecular_weights = {
+  [capabilities.formaldehydeMeasurement.NAME] = 30.031,
+}
+
+local conversion_tables = {
+  [units.PPM] = {
+    [units.PPM] = function(value) return utils.round(value) end,
+    [units.UGM3] = function(value, molecular_weight) return utils.round((value * molecular_weight * 10^3) / MGM3_PPM_CONVERSION_FACTOR) end,
+    [units.MGM3] = function(value, molecular_weight) return utils.round((value * molecular_weight) / MGM3_PPM_CONVERSION_FACTOR) end,
+  },
+  [units.PPB] = {
+    [units.PPM] = function(value) return utils.round(value/(10^3)) end
+  },
+  [units.PPT] = {
+    [units.PPM] = function(value) return utils.round(value/(10^6)) end
+  },
+  [units.MGM3] = {
+    [units.UGM3] = function(value) return utils.round(value * (10^3)) end,
+    [units.PPM] = function(value, molecular_weight) return utils.round((value * MGM3_PPM_CONVERSION_FACTOR) / molecular_weight) end,
+  },
+  [units.UGM3] = {
+    [units.UGM3] = function(value) return utils.round(value) end,
+    [units.PPM] = function(value, molecular_weight) return utils.round((value * MGM3_PPM_CONVERSION_FACTOR) / (molecular_weight * 10^3)) end,
+  },
+  [units.NGM3] = {
+    [units.UGM3] = function(value) return utils.round(value/(10^3)) end
+  },
+  [units.BQM3] = {
+    [units.PCIL] = function(value) return utils.round(value/37) end
+  },
+}
+
+local function unit_conversion(value, from_unit, to_unit, capability_name)
+  local conversion_function = conversion_tables[from_unit][to_unit]
+  if conversion_function == nil then
+    log.info_with( {hub_logs = true} , string.format("Unsupported unit conversion from %s to %s", unit_strings[from_unit], unit_strings[to_unit]))
+    return 1
```

Comment:
Why is a return value of 1 being used here? If the value can't be converted we should return an invalid value and handle it appropriately in the calling function so we aren't just emitting random 1's

Reply:
ok, we should be making these changes then in air-quality-sensor then. This part was just directly copied from there.

Reply:
Actually, looking at the problem more closely, I think this should be a separate PR. It should be fixed, but it is distinct enough and large enough that it should not be handled as an add-on here. This is especially true since it is handled equivalently in the air-quality-sensor subdriver

Reply:
FWIW I think we might want to port some of the other functions you have here to the AQS sub driver as well. I think the name building function you have made here is neater and we could modify the AQS driver to follow the same method when we make these changes regarding the return value as well